File: skills/current/x-research/lib/summary.ts
const SUMMARY_ROOT_FIELDS = [
  "requested_id",
  "requested_url",
  "handle",
  "query",
  "feed",
  "ranking_mode",
  "requested_count",
  "returned_count",
  "cursor",
  "has_more",
  "complete",
  "complete_reason",
  "provider",
  "official",
  "auth_mode",
  "source_url",
  "endpoint",
  "fetched_at",
  "provider_status",
] as const;

export function summaryAuthor(author: Record<string, unknown>): Record<string, unknown> {
  const summary: Record<string, unknown> = {};
  for (const key of ["id", "handle", "name", "url", "verified"] as const) {
    if (key in author && author[key] !== undefined) {
      summary[key] = author[key];
    }
  }
  return summary;
}

export function summaryPost(post: Record<string, unknown>): Record<string, unknown> {
  const summary: Record<string, unknown> = {};
  for (const key of ["id", "url", "text", "created_at", "lang", "quote_id", "reply_to_id"] as const) {
    if (key in post && post[key] !== undefined) {
      summary[key] = post[key];
    }
  }
  if ("author" in post && typeof post["author"] === "object" && post["author"] !== null) {
    summary["author"] = summaryAuthor(post["author"] as Record<string, unknown>);
  }
  return summary;
}

export function summaryPostValue(value: unknown): unknown {
  if (typeof value === "object" && value !== null && !Array.isArray(value)) {
    return summaryPost(value as Record<string, unknown>);
  }
  if (Array.isArray(value)) {
    return value
      .filter((item) => typeof item === "object" && item !== null && !Array.isArray(item))
      .map((item) => summaryPost(item as Record<string, unknown>));
  }
  return undefined;
}

export function summaryData(_command: string, data: Record<string, unknown>): Record<string, unknown> {
  const summary: Record<string, unknown> = {};
  for (const key of SUMMARY_ROOT_FIELDS) {
    if (key in data && data[key] !== undefined) {
      summary[key] = data[key];
    }
  }
  for (const key of ["post", "posts", "target", "thread", "replies"] as const) {
    if (!(key in data) || data[key] === undefined) {
      continue;
    }
    const projected = summaryPostValue(data[key]);
    if (projected !== undefined) {
      summary[key] = projected;
    }
  }
  if ("profile" in data && typeof data["profile"] === "object" && data["profile"] !== null) {
    summary["profile"] = summaryAuthor(data["profile"] as Record<string, unknown>);
  }
  return summary;
}

File: skills/current/x-research/lib/summary.py
"""Deterministic ``--summary`` projection of normalized command data."""

from typing import Any

from models import UNDEFINED

SUMMARY_ROOT_FIELDS = (
    "requested_id",
    "requested_url",
    "handle",
    "query",
    "feed",
    "ranking_mode",
    "requested_count",
    "returned_count",
    "cursor",
    "has_more",
    "complete",
    "complete_reason",
    "provider",
    "official",
    "auth_mode",
    "source_url",
    "endpoint",
    "fetched_at",
    "provider_status",
)


def summary_author(author: dict[str, Any]) -> dict[str, Any]:
    summary: dict[str, Any] = {}
    for key in ("id", "handle", "name", "url", "verified"):
        if key in author and author[key] is not UNDEFINED:
            summary[key] = author[key]
    return summary


def summary_post(post: dict[str, Any]) -> dict[str, Any]:
    summary: dict[str, Any] = {}
    for key in ("id", "url", "text", "created_at", "lang", "quote_id", "reply_to_id"):
        if key in post and post[key] is not UNDEFINED:
            summary[key] = post[key]
    author = post.get("author")
    if "author" in post and isinstance(author, dict):
        summary["author"] = summary_author(author)
    return summary


def summary_post_value(value: object) -> Any:
    if isinstance(value, dict):
        return summary_post(value)
    if isinstance(value, list):
        return [summary_post(item) for item in value if isinstance(item, dict)]
    return None


def summary_data(_command: str, data: dict[str, Any]) -> dict[str, Any]:
    summary: dict[str, Any] = {}
    for key in SUMMARY_ROOT_FIELDS:
        if key in data and data[key] is not UNDEFINED:
            summary[key] = data[key]
    for key in ("post", "posts", "target", "thread", "replies"):
        if key not in data or data[key] is UNDEFINED:
            continue
        projected = summary_post_value(data[key])
        if projected is not None:
            summary[key] = projected
    profile = data.get("profile")
    if "profile" in data and isinstance(profile, dict):
        summary["profile"] = summary_author(profile)
    return summary
